File: 04 Neo4J/backend/controllers/authController.js
const User = require('../models/userModel');
const bcrypt = require('bcrypt');
const jwt = require('jsonwebtoken');

exports.signup = async (req, res) => {
  const { username, password } = req.body;
  console.log(username);
  console.log(password);
  const hashedPassword = await bcrypt.hash(password, 10);
  try {
    const user = await User.create(username, hashedPassword);
    res.status(201).json({ message: 'User created successfully', user });
  } catch (error) {
    console.log(error);
    res.status(500).json({ error: 'Error creating user' });
  }
};

exports.login = async (req, res) => {
  const { username, password } = req.body;
  
  try {
    const userRecord = await User.findByUsername(username);
    if (!userRecord) {
      return res.status(404).json({ error: 'User not found' });
    }
    const validPassword = await bcrypt.compare(password, userRecord.properties.password);
    if (!validPassword) {
      return res.status(400).json({ error: 'Invalid password' });
    }
    const token = jwt.sign({ username: userRecord.properties.username }, 'secretkey', { expiresIn: '1h' });
    res.json({ message: 'Login successful', token });
  } catch (error) {
    res.status(500).json({ error: 'Error logging in' });
  }
};

File: 04 Neo4J/Python/backend/controllers/auth_controller.py
from flask import request, jsonify
from models.user_model import User
from werkzeug.security import generate_password_hash, check_password_hash
import jwt
import datetime

def signup():
    data = request.get_json()
    username = data['username']
    password = generate_password_hash(data['password'], method='sha256')
    try:
        User.create(username, password)
        return jsonify({'message': 'User created successfully'}), 201
    except Exception as e:
        return jsonify({'error': str(e)}), 500

def login():
    print('login')
    data = request.get_json()
    print(data)
    username = data['username']
    password = data['password']
    print("username: ", username)
    print("password: ", password)
    
    user_record = User.find_by_username(username)
    if not user_record:
        return jsonify({'error': 'User not found'}), 404
    
    if check_password_hash(user_record['password'], password):
        token = jwt.encode({'username': user_record['username'], 'exp': datetime.datetime.utcnow() + datetime.timedelta(hours=1)}, 'secretkey')
        return jsonify({'message': 'Login successful', 'token': token})
    else:
        return jsonify({'error': 'Invalid password'}), 400
